probe-rs/src/architecture/riscv/sequences/mod.rs

```diff
@@ -1,12 +1,167 @@
+    let old_ctrl_reg = interface.read_word_32(base | CTRL)?;
+    let old_user_reg = interface.read_word_32(base | USER)?;
+    let old_user1_reg = interface.read_word_32(base | USER1)?;
+
+    // ctrl register
+    const CTRL_WP: u32 = 1 << 21;
+
+    // user register
+    const USER_MISO: u32 = 1 << 28;
+    const USER_COMMAND: u32 = 1 << 31;
+
+    // user2 register
+    const USER_COMMAND_BITLEN: u32 = 28;
+
+    // miso dlen register
+    const MISO_BITLEN: u32 = 0;
+
+    // cmd register
+    const USER_CMD: u32 = 1 << 18;
+
+    interface.write_word_32(base | CTRL, old_ctrl_reg | CTRL_WP)?;
+    interface.write_word_32(base | USER, old_user_reg | USER_COMMAND | USER_MISO)?;
+    interface.write_word_32(base | USER1, 0)?;
+    interface.write_word_32(base | USER2, (7 << USER_COMMAND_BITLEN) | command as u32)?;
+    interface.write_word_32(
+        base | MISO_DLEN,
+        (miso_bits.saturating_sub(1)) << MISO_BITLEN,
+    )?;
+    interface.write_word_32(base | DATA_BUF_0, 0)?;
+
+    // Execute read
+    interface.write_word_32(base | CMD, USER_CMD)?;
+    while interface.read_word_32(base | CMD)? & USER_CMD != 0 {}
+
+    // Read result
+    let value = interface.read_word_32(base | DATA_BUF_0)?;
+
+    // Restore registers
+    interface.write_word_32(base | CTRL, old_ctrl_reg)?;
+    interface.write_word_32(base | USER, old_user_reg)?;
+    interface.write_word_32(base | USER1, old_user1_reg)?;
+
+    Ok(value)
+}
+
+fn esp32_attach_flash(
+    interface: &mut RiscvCommunicationInterface,
+    stack_pointer: u32,
+    attach_fn: u32,
+) -> Result<(), crate::Error> {
+    // Set a valid-ish stack pointer
+    interface.abstract_cmd_register_write(SP, stack_pointer)?;
+
+    // esp_rom_spiflash_attach(0, false)
+    interface.schedule_setup_program_buffer(&[
+        assembly::addi(0, 10, 0),                       // c.li a0, zero
+        assembly::addi(0, 11, 0),                       // c.li a1, zero
+        assembly::lui(5, (attach_fn >> 12) as i32),     // lui x5, ...
+        assembly::jarl(5, 1, attach_fn as i32 & 0xFFF), // jarl ra, x5, ...
+    ])?;
+
+    // Actually call the function
+    let mut execute_progbuf = AccessRegisterCommand(0);
+    execute_progbuf.set_cmd_type(0);
+    execute_progbuf.set_transfer(false);
+    execute_progbuf.set_write(false);
+    execute_progbuf.set_aarsize(RiscvBusAccess::A32);
+    execute_progbuf.set_postexec(true);
+    execute_progbuf.set_regno(SP.id.0 as u32);
+
+    interface.schedule_write_dm_register(execute_progbuf)?;
+    interface.execute()?;
+
+    Ok(())
+}
+
+fn esp32_detect_flash_size(
+    interface: &mut RiscvCommunicationInterface,
+    spiflash_addr: u32,
+) -> Result<Option<usize>, crate::Error> {
+    const RDID: u8 = 0x9F;
+    let value = esp32_execute_flash_command(interface, spiflash_addr, RDID, 24)?;
+
+    Ok(decode_flash_size(value))
+}
+
+fn decode_flash_size(value: u32) -> Option<usize> {
+    let [manufacturer, memory_type, capacity, _] = value.to_le_bytes();
```

Comment:
I think the jep106 crate can help us here. I think we already depend on it too, if we wanted to print manufacturer details etc.

Reply:
Good point, but I'm not sure if we need it right now.